onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -25,72 +25,140 @@ size_t ComputeOffset(const std::vector<int64_t>& index, const std::vector<size_t
+  const std::string* end = source + num_elts_in_block;
+  std::copy(source, end, target);
+}
+
 // DoTranspose: copies source tensor to target, transposing elements.
 // The stride vector indicates the transposition.
-template <typename T>
 static void DoTransposeImpl(int64_t num_axes, const std::vector<int64_t>& target_dims,
                             size_t num_blocks, size_t num_elts_in_block, const std::vector<size_t>& stride,
-                            const T* source, T* target) {
-  size_t blocksize = num_elts_in_block * sizeof(float);
+                            const uint8_t* source, uint8_t* target, size_t element_size) {
+  size_t blocksize = num_elts_in_block * element_size;
+  // index used to iterate over target iteration-space
+  std::vector<int64_t> target_index(num_axes, 0);
+  for (size_t i = 0; i < num_blocks; ++i) {
+    // convert target_index into an offset in source data
+    size_t source_offset = ComputeOffset(target_index, stride, num_axes);
+
+    // copy
+    memcpy(target, source + source_offset * element_size, blocksize);
+
+    // increment target_index:
+    IncrementIndex(target_index, target_dims, num_axes);
+    target += blocksize;
+  }
+}
+
+static void DoTransposeImpl(int64_t num_axes, const std::vector<int64_t>& target_dims,
+                            size_t num_blocks, size_t num_elts_in_block, const std::vector<size_t>& stride,
+                            const std::string* source, std::string* target) {
   // index used to iterate over target iteration-space
   std::vector<int64_t> target_index(num_axes, 0);
   for (size_t i = 0; i < num_blocks; ++i) {
     // convert target_index into an offset in source data
     size_t source_offset = ComputeOffset(target_index, stride, num_axes);
 
     // copy
-    memcpy(target, source + source_offset, blocksize);
+    DoTransposeSingleBlock(num_elts_in_block, source + source_offset, target);
 
     // increment target_index:
     IncrementIndex(target_index, target_dims, num_axes);
     target += num_elts_in_block;
   }
 }
 
+template <class T>
+inline void CopyPrim(const uint8_t* source, uint8_t* target) {
+  *reinterpret_cast<T*>(target) = *reinterpret_cast<const T*>(source);
+}
+
+static inline void CopyPrimitiveElement(const uint8_t* source, uint8_t* target, size_t element_size) {
+  // Make sure some assumptions hold
+  static_assert(sizeof(MLFloat16) == sizeof(uint16_t), "MLFloat16 not covered");
+  static_assert(sizeof(BFloat16) == sizeof(uint16_t), "BFloat16 not covered");
+  switch (element_size) {
+    case sizeof(uint64_t):
+      CopyPrim<uint64_t>(source, target);
+      break;
+    case sizeof(uint32_t):
+      CopyPrim<uint32_t>(source, target);
+      break;
+    case sizeof(uint16_t):
+      CopyPrim<uint16_t>(source, target);
+      break;
+    case sizeof(uint8_t):
+      *target = *source;
+      break;
+    default:
+      assert(false);
+  }
+}
```

Comment:
Thanks for making a generic kernel for Tranpose! Wrt. the implementation, I think you may try Eigen::Tensor to see if perf is better.

Reply:
Any reference to what you mean here?

Reply:
So that would mean going back to templated implementation with code bloating? We need to make up our mind here first. + separate std::string code.

---
In reply to: [263603562](https://github.com/Microsoft/onnxruntime/pull/561#discussion_r263603562) [](ancestors = 263603562)

Reply:
You only need 5 templates for transpose: uint8/uint16/uint32/uint64/string. The benefit of using Eigen is that compiler might have better chance using vector instructions.

---
In reply to: [263880691](https://github.com/Microsoft/onnxruntime/pull/561#discussion_r263880691) [](ancestors = 263880691,263603562)

Reply:
5 templates + 1 generic implementation + 1 string implementation, that's 7. Today at the scrum , reducing the binary size was marked as one of the priorities. I would like to check this in as it is and work on perf only if necessary.

Reply:
Let's not have the template solution right now, we may take it if seeing perf concern, as we're having high pressure on binary size reduction.